src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -18936,5 +18996,1565 @@ interface I1
+            Program.M(s);
+            System.Console.Write(3);
+        }
+        else
+        {
+            System.Console.Write(4);
+        }
+    }
+}
+
+ref struct S
+{
+}
+
+class Program
+{
+    static void Main()
+    {
+        Helper1<S>.Test1(new S());
+        Helper1<int>.Test1(1);
+        Helper2.Test2(new S());
+    }
+
+    public static void M<T>(T t) where T : allows ref struct {}
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "113" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).
+            VerifyDiagnostics();
+
+            verifier.VerifyIL("Helper1<T>.Test1(T)",
+@"
+{
+  // Code size       15 (0xf)
+  .maxstack  1
+  .locals init (T V_0) //t
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloc.0
+  IL_0003:  call       ""void Program.M<T>(T)""
+  IL_0008:  ldc.i4.1
+  IL_0009:  call       ""void System.Console.Write(int)""
+  IL_000e:  ret
+}
+");
+
+            verifier.VerifyIL("Helper2.Test2(S)",
+@"
+{
+  // Code size       15 (0xf)
+  .maxstack  1
+  .locals init (S V_0) //s
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloc.0
+  IL_0003:  call       ""void Program.M<S>(S)""
+  IL_0008:  ldc.i4.3
+  IL_0009:  call       ""void System.Console.Write(int)""
+  IL_000e:  ret
+}
+");
+        }
+
+        [Fact]
+        public void IsPattern_05()
+        {
+            var src = @"
+class Helper1<T, U>
+    where T : allows ref struct
+{
+    public static void Test1(T h1)
+    {
+        if (h1 is U u)
+        {
+        }
+    }
+}
+
+class Helper2<U>
+{
+    public static void Test2(S h2)
+    {
+        if (h2 is U u)
+        {
+        }
+    }
+}
+
+ref struct S : I1
+{
+}
+
+interface I1
+{
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (7,19): error CS8121: An expression of type 'T' cannot be handled by a pattern of type 'U'.
+                //         if (h1 is U u)
+                Diagnostic(ErrorCode.ERR_PatternWrongType, "U").WithArguments("T", "U").WithLocation(7, 19),
+                // (17,19): error CS8121: An expression of type 'S' cannot be handled by a pattern of type 'U'.
+                //         if (h2 is U u)
+                Diagnostic(ErrorCode.ERR_PatternWrongType, "U").WithArguments("S", "U").WithLocation(17, 19)
+                );
+        }
+
+        [Fact]
+        public void IsPattern_06()
+        {
+            var src = @"
+class Helper1<T, U>
+    where T : allows ref struct
+    where U : allows ref struct
+{
+    public static void Test1(T h1)
+    {
+        if (h1 is U u)
+        {
+        }
+    }
+}
```

Comment:
Consider testing:
```csharp
class Helper1<T, U>
    where T : allows ref struct
    where U : T, allows ref struct
{
    public static void Test1(T h1)
    {
        if (h1 is U u)
        {
        }
    }
    public static void Test2(U h2)
    {
        if (h2 is T t)
        {
        }
    }
}
```

Reply:
> Consider testing:

I will add the following test in the next PR:
```
        [Fact]
        public void IsPattern_10()
        {
            var src = @"
class Helper1<T, U>
    where T : allows ref struct
    where U : T, allows ref struct
{
    public static void Test1(T h1)
    {
        if (h1 is U u)
        {
        }
    }
    public static void Test2(U h2)
    {
        if (h2 is T t)
        {
        }
    }
}
";

            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
            comp.VerifyDiagnostics(
                // (8,19): error CS8121: An expression of type 'T' cannot be handled by a pattern of type 'U'.
                //         if (h1 is U u)
                Diagnostic(ErrorCode.ERR_PatternWrongType, "U").WithArguments("T", "U").WithLocation(8, 19),
                // (14,19): error CS8121: An expression of type 'U' cannot be handled by a pattern of type 'T'.
                //         if (h2 is T t)
                Diagnostic(ErrorCode.ERR_PatternWrongType, "T").WithArguments("U", "T").WithLocation(14, 19)
                );
        }
```